File: webapp/walk_ticket.py
"""Short-lived walk tickets: the web server vouches for the logged-in account name to the multiplayer server.

Format: base64url(JSON {"name", "exp"}) + "." + hex HMAC-SHA256 of that base64 text, keyed with
WALK_TICKET_SECRET shared by both servers. The multiplayer server takes the player's name only from a valid ticket.
"""
import base64
import hashlib
import hmac
import json
import time

TTL_SECONDS = 60


def make_ticket(name, secret, now=None, ttl=TTL_SECONDS):
    payload = base64.urlsafe_b64encode(json.dumps({'name': name, 'exp': int((now or time.time()) + ttl)}, ensure_ascii=False).encode()).decode().rstrip('=')
    signature = hmac.new(secret.encode(), payload.encode(), hashlib.sha256).hexdigest()
    return f'{payload}.{signature}'

File: multiplayer/walk_ticket.js
import { createHmac, timingSafeEqual } from 'node:crypto';
import { normalizePlayerName } from '../webapp/static/player_name.js';

// Walk tickets are issued by the web server for the logged-in account (webapp/walk_ticket.py):
// base64url(JSON {name, exp}) + "." + hex HMAC-SHA256 of the base64 text with the shared secret.
export function makeTicket(name, secret, now = Date.now() / 1000, ttl = 60) {
  const payload = Buffer.from(JSON.stringify({ name, exp: Math.floor(now + ttl) })).toString('base64url');
  return `${payload}.${createHmac('sha256', secret).update(payload).digest('hex')}`;
}

export function verifyTicket(ticket, secret, now = Date.now() / 1000) {
  if (!secret || typeof ticket !== 'string' || ticket.length > 600) return null;
  const [payload, signature, extra] = ticket.split('.');
  if (!payload || !signature || extra !== undefined || !/^[0-9a-f]{64}$/.test(signature)) return null;
  const expected = createHmac('sha256', secret).update(payload).digest();
  if (!timingSafeEqual(expected, Buffer.from(signature, 'hex'))) return null;
  try {
    const { name, exp } = JSON.parse(Buffer.from(payload, 'base64url').toString('utf8'));
    if (!Number.isFinite(exp) || exp < now) return null;
    return normalizePlayerName(name);
  } catch { return null; }
}
